otherlibs/systhreads/Makefile

```diff
@@ -13,96 +13,127 @@
+	$(AR) rc $@ $^
+else # Windows
+	$(MKLIB) -o $(LIBNAME)nat $^ $(LDOPTS)
+endif
 
-st_stubs_n.o: st_stubs.c st_posix.h
-	$(NATIVECC) -I../../asmrun -I../../byterun $(NATIVECCCOMPOPTS) \
-	            $(SHAREDCCCOMPOPTS) -DNATIVE_CODE -DTARGET_$(ARCH) \
-	            -DMODEL_$(MODEL) -DSYS_$(SYSTEM) -c st_stubs.c
-	mv st_stubs.o st_stubs_n.o
+$(LIBNAME).cma: $(THREADS_BCOBJS)
+ifeq "$(UNIX_OR_WIN32)" "unix"
+	$(MKLIB) -o $(LIBNAME) -ocamlc '$(CAMLC)' -cclib -lunix $(PTHREAD_CAML_LINK) $^
+else # Windows
+	$(MKLIB) -o $(LIBNAME) -ocamlc "$(CAMLC)" -linkall $^
+endif
 
-threads.cma: $(THREAD_OBJS)
-	$(MKLIB) -ocamlc '$(CAMLC)' -o threads $(THREAD_OBJS) \
-	  -cclib -lunix $(PTHREAD_CAML_LINK)
+# TODO: figure out why linkall is used only for Windows
 
 # See remark above: force static linking of libthreadsnat.a
-threads.cmxa: $(THREAD_OBJS:.cmo=.cmx)
-	$(CAMLOPT) -a -o threads.cmxa $(THREAD_OBJS:.cmo=.cmx) \
-	  -cclib -lthreadsnat $(PTHREAD_CAML_LINK)
+$(LIBNAME).cmxa: $(THREADS_NCOBJS)
+	$(CAMLOPT) -a -o $@ $^ -cclib -lthreadsnat $(PTHREAD_CAML_LINK)
 
 # Note: I removed "-cclib -lunix" from the line above.
 # Indeed, if we link threads.cmxa, then we must also link unix.cmxa,
 # which itself will pass -lunix to the C linker.  It seems more
 # modular to me this way. -- Alain
 
+$(LIBNAME).cmxs: $(LIBNAME).cmxa lib$(LIBNAME)nat.$(A)
+	$(CAMLOPT) -shared -o $@ -I . $< -linkall
+
+st_stubs_b.$(O): st_stubs.c $(HEADER)
+	$(BYTECC) -I$(ROOTDIR)/byterun $(BYTECCCOMPOPTS) $(SHAREDCCCOMPOPTS) -c $<
+	mv st_stubs.$(O) $@
 
-$(THREAD_OBJS:.cmo=.cmx): ../../ocamlopt
+st_stubs_n.$(O): st_stubs.c $(HEADER)
+	$(NATIVECC) -I$(ROOTDIR)/asmrun -I$(ROOTDIR)/byterun $(NATIVECCCOMPOPTS) \
+	  $(SHAREDCCCOMPOPTS) -DNATIVE_CODE -DTARGET_$(ARCH) \
+	  -DMODEL_$(MODEL) -DSYS_$(SYSTEM) -c $<
+	mv st_stubs.$(O) $@
+
+$(THREAD_NCOBJS): $(ROOTDIR)/ocamlopt
 
 partialclean:
 	rm -f *.cm*
 
 clean: partialclean
-	rm -f *.o *.a *.so
+	rm -f *$(EXT_DLL) *.$(A) *.$(O)
 
 INSTALL_LIBDIR=$(DESTDIR)$(LIBDIR)
+THREADS_LIBDIR=$(INSTALL_LIBDIR)/threads
 INSTALL_STUBLIBDIR=$(DESTDIR)$(STUBLIBDIR)
 
 install:
-	if test -f dllthreads.so; then \
-	  cp dllthreads.so $(INSTALL_STUBLIBDIR)/dllthreads.so; fi
-	cp libthreads.a $(INSTALL_LIBDIR)/libthreads.a
-	cd $(INSTALL_LIBDIR); $(RANLIB) libthreads.a
-	if test -d $(INSTALL_LIBDIR)/threads; then :; \
-	  else mkdir $(INSTALL_LIBDIR)/threads; fi
-	cp $(THREAD_OBJS:.cmo=.cmi) threads.cma $(INSTALL_LIBDIR)/threads
-	rm -f $(INSTALL_LIBDIR)/threads/stdlib.cma
-	cp thread.mli mutex.mli condition.mli event.mli threadUnix.mli \
-	   $(INSTALL_LIBDIR)
-	cp threads.h $(INSTALL_LIBDIR)/caml/threads.h
+	if test -f dllthreads$(EXT_DLL); then \
+	  cp dllthreads$(EXT_DLL) $(INSTALL_STUBLIBDIR); fi
+	cp libthreads.$(A) $(INSTALL_LIBDIR)
+ifeq "$(UNIX_OR_WIN32)" "unix"
+	cd $(INSTALL_LIBDIR); $(RANLIB) libthreads.$(A)
+endif
+	mkdir -p "$(THREADS_LIBDIR)"
+	cp $(CMIFILES) threads.cma "$(THREADS_LIBDIR)"
+	rm -f "$(THREADS_LIBDIR)/stdlib.cma"
```

Comment:
Alain Frisch (2016/08/31 08:46 -0700):
> Yes:
> 
>     rm -f "$(THREADS_LIBDIR)/stdlib.cma"
> 
> I don't know what is this stdlib.cma under `$(THREADS_LIBDIR)`.

Me neither. Is it possible that, for older versions of OCaml, the stdlib
came in two flavours depending on whether one compiled with or without
threads? And then when newer versions arrived they added this command to
remove the thread-safe version of the stdlib which became obsolete?

Is that possible?

Perhaps this command can now be removed?